### ISSUE TITLE
Hang in exit in TRAM test on gni-crayxc-smp

### ISSUE THREAD

**Issue by stwhite91** [UIUC-PPL/charm#1329]
*Original issue: https://charm.cs.illinois.edu/redmine/issues/1329*

---
tests/charm++/streamingAllToAll/ appears to be hanging every night in autobuild on gni-crayxc-smp then timing out:

http://charm.cs.uiuc.edu/autobuild/cur/gni-crayxc-smp.txt

```
Entering directory `/scratch1/scratchdirs/acun/autobuild/gni-crayxc-smp/charm/gni-crayxc-smp/tests/charm++/streamingAllToAll'
srun -n 4 -c 2 ./ataTest +CmiNoProcForComThread
Charm++> Running on Gemini (GNI) with 4 processes
Charm++> static SMSG
Charm++> SMSG memory: 19.8KB
Charm++> memory pool init block size: 8MB, total memory pool limit 0MB (0 means no limit)
Charm++> memory pool registered memory limit: 200000MB, send limit: 100000MB
Charm++> only comm thread send/recv messages
Charm++> Cray TLB page size: 8192K
Charm++> Running in SMP mode: numNodes 4,  1 worker threads per process
Charm++> The comm. thread both sends and receives messages
Converse/Charm++ Commit ID: 0e19d77
Charm++> Note: The option +CmiNoProcForComThread has been superseded by +CmiSleepOnIdle
Warning> using Isomalloc in SMP mode, you may need to run with '+isomalloc_sync'.
CharmLB> Load balancer assumes all CPUs are same.
Charm++> Running on 1 unique compute nodes (48-way SMP).
size of envelope: 80

TEST 1: Using 2D TRAM Topology: 4 1
Elapsed time for all-to-all of       32 bytes sent in      1  iteration of 32 bytes each (    using TRAM): 0.019991 seconds
Elapsed time for all-to-all of       64 bytes sent in      2 iterations of 32 bytes each (    using TRAM): 0.000056 seconds
Elapsed time for all-to-all of      128 bytes sent in      4 iterations of 32 bytes each (    using TRAM): 0.000043 seconds
Elapsed time for all-to-all of      256 bytes sent in      8 iterations of 32 bytes each (    using TRAM): 0.000040 seconds
Elapsed time for all-to-all of      512 bytes sent in     16 iterations of 32 bytes each (    using TRAM): 0.000040 seconds
Elapsed time for all-to-all of     1024 bytes sent in     32 iterations of 32 bytes each (    using TRAM): 0.000193 seconds
Elapsed time for all-to-all of     2048 bytes sent in     64 iterations of 32 bytes each (    using TRAM): 0.000050 seconds
Elapsed time for all-to-all of     4096 bytes sent in    128 iterations of 32 bytes each (    using TRAM): 0.003155 seconds
Elapsed time for all-to-all of     8192 bytes sent in    256 iterations of 32 bytes each (    using TRAM): 0.000087 seconds
Elapsed time for all-to-all of    16384 bytes sent in    512 iterations of 32 bytes each (    using TRAM): 0.000136 seconds

TEST 2: Using point to point sends
Elapsed time for all-to-all of       32 bytes sent in      1  iteration of 32 bytes each (not using TRAM): 0.000032 seconds
Elapsed time for all-to-all of       64 bytes sent in      2 iterations of 32 bytes each (not using TRAM): 0.000051 seconds
Elapsed time for all-to-all of      128 bytes sent in      4 iterations of 32 bytes each (not using TRAM): 0.000060 seconds
Elapsed time for all-to-all of      256 bytes sent in      8 iterations of 32 bytes each (not using TRAM): 0.000075 seconds
Elapsed time for all-to-all of      512 bytes sent in     16 iterations of 32 bytes each (not using TRAM): 0.000130 seconds
Elapsed time for all-to-all of     1024 bytes sent in     32 iterations of 32 bytes each (not using TRAM): 0.000199 seconds
Elapsed time for all-to-all of     2048 bytes sent in     64 iterations of 32 bytes each (not using TRAM): 0.000381 seconds
Elapsed time for all-to-all of     4096 bytes sent in    128 iterations of 32 bytes each (not using TRAM): 0.000838 seconds
Elapsed time for all-to-all of     8192 bytes sent in    256 iterations of 32 bytes each (not using TRAM): 0.001428 seconds
Elapsed time for all-to-all of    16384 bytes sent in    512 iterations of 32 bytes each (not using TRAM): 0.002528 seconds
[Partition 0][Node 0] End of program
slurmstepd: error: *** STEP 2982328.38 ON nid00546 CANCELLED AT 2016-12-12T05:08:20 DUE TO TIME LIMIT ***
slurmstepd: error: *** JOB 2982328 ON nid00546 CANCELLED AT 2016-12-12T05:08:20 DUE TO TIME LIMIT ***

```

**Comment by stwhite91** [comment]
*Original date: 2016-12-14 21:38:42*

---
Assigning to Karthik because he is looking at optimizations for TRAM in SMP mode.

Also note that this hang is nondeterministic, it doesn't happen every time.

**Comment by stwhite91** [comment]
*Original date: 2017-01-26 18:44:08*

---
Core group would like to see some progress on this issue. Do you have access to Edison to reproduce this?

**Comment by karthiksenthil** [comment]
*Original date: 2017-02-01 02:32:27*

---
The bug does not happen in 'debug' build of Charm++ on Edison.

I am currently trying to find the exact location of the bug. The program flow reaches `CkExit()` invocation and there is no QD involved in the program or TRAM library. One possible suspect could be the `CthYield()` that is explicitly called in the non-TRAM test case whenever counter reaches 1024.

**Comment by PhilMiller** [comment]
*Original date: 2017-02-09 22:06:28*

---
Any progress on this? If you need help, please ask for it.

**Comment by karthiksenthil** [comment]
*Original date: 2017-02-10 17:22:06*

---
I have isolated the bug and it occurs in the "directSends" version of the test. The "usingTram" part is clean. As suspected it is mostly occurring due to some race condition or mistakes in invocations made to `CthYield()`. I am looking into this right now and will update as soon as I find the fix.

**Comment by karthiksenthil** [comment]
*Original date: 2017-02-11 09:06:39*

---
Some notes on the experiments I performed today for this bug:
https://charm.cs.illinois.edu/newTms/tasks/1753

**Comment by stwhite91** [comment]
*Original date: 2017-02-13 22:40:32*

---
The hang is not related to TRAM, it just happens in the TRAM test.

The real issue seems to be somewhere in our exit sequence on gni-smp builds.

**Comment by karthiksenthil** [comment]
*Original date: 2017-02-13 23:26:47*

---
The hang does not happen when using run options like `+p4 ++ppn4`. However, for `+p8 ++ppn4` the program hangs.

**Comment by PhilMiller** [comment]
*Original date: 2017-02-21 18:23:09*

---
Fix in progress here:
~~https://charm.cs.illinois.edu/gerrit/#/c/2236/~~ https://github.com/UIUC-PPL/charm/commit/ae20bd242ddc64239a8d9a9867be94b8365cec51

**Comment by karthiksenthil** [comment]
*Original date: 2017-02-26 17:23:31*

---
Debugged the test case with various setups. Following are some notes from the experiments :
* The hang is not related to TRAM as such. I commented out all code in the test case related to TRAM and program still hangs
* It is also not related to threaded entry methods(`Participant::communicate`) and explicit invocations to `CthYield()`. I ran multiple tests with code related to this removed, but the program still hangs.
* Premature call to `CkExit()` is possible in the test case. The chares in the `allToAllGroup` (a group chare) contribute to exit after messaging all other chares in the same array. It is possible that all these chares process the contribute call before processing messages from other chares, which leads to early `CkExit()`
* However is this premature exit an issue since these messages are not significant for the test program? If yes, why doesn't the bug reproduce in other builds on other machines?
* This could also be a bug related to Charm++ internal exit sequence handling for group chares (particularly for gni layer?)
